packages/cpp/ArmoniK.Api.Client/source/channel/ChannelFactory.cpp

```diff
@@ -0,0 +1,165 @@
+AwEAAaAAMA0GCSqGSIb3DQEBCwUAA4ICAQCr75dBYjypzqDqQ6TiKWuYO8rq6TIh
+pdZHw5ystwvD6sn+tPbc7iNbnvDF6GeTgMdKAuwNz0YJMZq9v39hZzTCyMqRLNlT
+TU3kYaTWGHDK0HE2O3pHKppHAc2YbAsSxuS8KMHx0wW0abVHiEeudc/nULJppX1/
+ObouzLGSJJwZctXEzk/Ye7bD1sneSqVnrdFD1IOBVQVRGoJznAt7WWxvGk9LPW51
++MybzTilL4rk5+ezA4UCIMrQCDwZcI+UCcKqDajDz+7kn81f1K4g1G6dTh+M8qIV
+lx6/Bfy3P6DHF1ww0i/hRQht1O9cyUo3mDZzAq20OsIDvkhjNGma/IEbkZ9z0P5C
+/5YwAW+GuwG2GrD016y5OjZVrAG/KIfyS6FLQfgN/ww5Y9tK6vO5XkelED7zNPrq
+em1zkId2H0Az5dIC2OpnAg3+NuGrehfIXziiY+8MGIivqI/Rulnv7m2l2vjHi66K
+GztDm5ohMdfjitFIfPDFYPMH7KES4vivic8zlq9FJYNp8tUYEBR1wW7W03IJPm6e
+pUwvXHPjId/qBjlBixZt2ZqC8X4S95wAfVjtS3O33Zsm4oevwlvywfYIK8nTG5SD
+bDCNVTg3w/OQLQQdWUl6FunmYinukBgmqnsJnwgrhzBENbmgbgfOZZWGtG5ODENb
+wc+KqiSg9c9iqA==
+-----END CERTIFICATE REQUEST-----)";
+
+std::string get_key(const absl::string_view &path) {
+  std::ifstream file(path.data(), std::ios::in | std::ios::binary);
+  if (file.is_open()) {
+    std::ostringstream sstr;
+    sstr << file.rdbuf();
+    return sstr.str();
+  } else {
+    return {};
+  }
+}
+
+bool initialize_protocol_endpoint(const common::options::ControlPlane &controlPlane, std::string &endpoint) {
+  absl::string_view endpoint_view = controlPlane.getEndpoint();
+  const auto delim = endpoint_view.find("://");
+  if (delim != absl::string_view::npos) {
+    const auto tmp = endpoint_view.substr(delim + 3);
+    endpoint_view = endpoint_view.substr(0, delim);
+    endpoint = {tmp.cbegin(), tmp.cend()};
+  }
+  return endpoint_view.back() == 's' || endpoint_view.back() == 'S';
+}
+
+std::shared_ptr<CertificateProviderInterface> create_certificate_provider(const std::string &rootCertificate,
+                                                                          const std::string &userPublicPem,
+                                                                          const std::string &userPrivatePem) {
+  if (rootCertificate.empty()) {
+    return std::make_shared<StaticDataCertificateProvider>(
+        std::vector<IdentityKeyCertPair>{IdentityKeyCertPair{userPrivatePem, userPublicPem}});
+  } else if (userPrivatePem.empty() || userPublicPem.empty()) {
+    return std::make_shared<StaticDataCertificateProvider>(rootCertificate);
+  } else {
+    return std::make_shared<StaticDataCertificateProvider>(
+        rootCertificate, std::vector<IdentityKeyCertPair>{IdentityKeyCertPair{userPrivatePem, userPublicPem}});
+  }
+}
+
+std::shared_ptr<grpc::Channel> ChannelFactory::create_channel() {
+  auto channel = grpc::CreateCustomChannel(endpoint_, credentials_, common::utils::getChannelArguments(configuration_));
+  logger_.log(common::logger::Level::Debug, "Created new channel ");
+
+  if (channel != nullptr) {
+    if (ShutdownOnFailure(channel)) {
+      logger_.log(common::logger::Level::Debug, "Shutdown unhealthy channel");
+    } else {
+      logger_.log(common::logger::Level::Debug, "Valid channel");
+    }
+  }
+  return channel;
+}

```

Comment:
Why ? You just created it so it's not null. Shutdown on failure is used in channel pools, but not for this kind of factory

Reply:
I used the "Shutdown on failure" because when debugging I noticed that the created channel can be in failure that's why I added the check after the creation.

Reply:
I think the comment was more about `if (channel != nullptr)`